packages/jest-haste-map/src/index.ts

```diff
@@ -115,6 +117,7 @@ export type {default as FS} from './HasteFS';
 export type {ChangeEvent, HasteMap as HasteMapObject} from './types';
 
 const CHANGE_INTERVAL = 30;
+const DEFAULT_WORKER_TIMEOUT = 3000;
```

Comment:
Crawling can potentially take way longer than this. I'd default to at least 30 sec, if not a full minute

Reply:
Alright! Any tips for publicly available resources to use for building up a test environment for these kind of things? E.g. public repos with A LOT of jest tests?

PS: [This](https://youtu.be/3YDiloj8_d0?t=1155) just popped into my mind, @cpojer  😄

Reply:
Crawling is not done in the worker processes though.

The workers are pluggable, so users could theoretically do anything they want in those workers. While that shouldn't be something that people generally do, they would need to be able to customize this timeout though.

Reply:
> Crawling is not done in the worker processes though.

Indeed, building the haste map using the workers happens after file system crawling.

Do you have an idea of how long this could be expected to take on a very large codebase without watchman?

I went with 3000 ms as a starting point for discussion here with the thought that if it's 'too long' it negates some of the point of having the timeout; giving feedback in a reasonable time that something went wrong and exiting.

> The workers are pluggable, so users could theoretically do anything they want in those workers. While that shouldn't be something that people generally do, they would need to be able to customize this timeout though.

Agreed; that's why I made the timeout mechanism in `jest-haste-map` and configurable via `haste.workerTimeout` in the project config when used via Jest, or the config object passed in when instantiating HasteMap. Meaning it should be accessible for configuration also if someone were to use `jest-haste-map` alone in some other project.

Does that seem sufficient?

Reply:
If you don't already have numbers on it, I'm thinking that I could make a script to duplicate all the jest packages in increments and benchmark the code in question in between. Running that on a selection of environments with different hardware configurations and preexisting loads ought to serve as a decent enough analogue for how it scales and take the guesswork out of it. 🤔

Reply:
Oh sorry I missed the timeout config.

You can create a folder with thousands of JS files and run Jest with `--no-cache` for testing.

I think 3 seconds per file analysis is fine. The only thing that worries me is that it is not entirely up to Jest, like if there is file system read contention, then it will timeout here even though it would eventually complete. For that reason it probably makes sense to bump the timeout to 10 seconds at least (as the intent here is to discover serious errors, not actually slow running processes, right?).

Reply:
Alright, thanks - I'm a busy traveling the next days, but will put something together when I can! 👍  

I agree that especially seeing how this handles under resource constraints would be prudent. And that's right, the purpose is to (with minimal changes) discover serious errors that could otherwise go unhandled. To that end, 10 seconds seems reasonable 🙂

Reply:
Hey guys! Just wanted to apologise for not following up on this. My off-the-grid holiday in Norway became even more disconnected than planned. Anyhoo... I'm back to civilisation and will soon be back in action. 🙂  

Well... once I've chewed through two months worth of email and notifications. Oh my! :sweat_smile: